src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -674,3 +682,430 @@ def _standard_metadata(self, name: str) -> V1ObjectMeta:
+        clumsy, and the created Ingress is a good proxy.
+
+        If we find a broken fileserver, we delete all its objects.  In
+        steady-state operations, this will happen when a fileserver Pod has
+        seen no operations for its timeout period; the Pod will exit and
+        the reconciliation task will note that the Job has no active Pods
+        and take action to clean it up.
+        """
+        obj_name = f"{username}-fs"
+        self._logger.debug(f"Checking whether {username} has fileserver")
+        try:
+            self._logger.debug(f"Checking job for {username}")
+            job = await self.batch_api.read_namespaced_job(obj_name, namespace)
+        except ApiException as e:
+            self._logger.info(f"Job {obj_name} for {username} not found.")
+            if e.status == 404:
+                return False
+            raise KubernetesError.from_exception(
+                "Error reading job",
+                e,
+                namespace=namespace,
+                name=obj_name,
+            ) from e
+        if job.status is None or job.status.active < 1:
+            self._logger.info(
+                f"Job {obj_name} has no active pods; "
+                + f"terminating fileserver for {username}"
+            )
+            return False
+        try:
+            self._logger.debug(f"Checking service for {username}")
+            job = await self.api.read_namespaced_service(obj_name, namespace)
+        except ApiException as e:
+            self._logger.info(f"Service {obj_name} for {username} not found.")
+            if e.status == 404:
+                return False
+            raise KubernetesError.from_exception(
+                "Error reading service",
+                e,
+                namespace=namespace,
+                name=obj_name,
+            ) from e
+        try:
+            self._logger.debug(f"Checking ingress for {username}")
+            await self.networking_api.read_namespaced_ingress(
+                obj_name, namespace
+            )
+        except ApiException as e:
+            self._logger.info(f"Ingress {obj_name} for {username} not found.")
+            if e.status == 404:
+                return False
+            raise KubernetesError.from_exception(
+                "Error reading service",
+                e,
+                namespace=namespace,
+                name=obj_name,
+            ) from e
+        return True
+
+    async def remove_fileserver(self, username: str, namespace: str) -> None:
+        """Remove the set of fileserver objects for a user.  It doesn't
+        return until the objects are no longer present in Kubernetes.
+        """
+        with FILESERVER_LOCK:
+            await self._delete_fileserver_job(username, namespace)
+            await self._delete_fileserver_service(username, namespace)
+            await self._delete_fileserver_gafaelfawringress(
+                username, namespace
+            )
```

Comment:
As mentioned elsewhere, this logic should move to the service layer; it's too much business logic for the storage layer. But when doing that, the ingress should be deleted first, then the service, then the job.

Reply:
Will do with regards to the ordering.

Reply:
Done (with respect to the ordering).  Service layer migration deferred till later.